permissions/src/main/java/com/google/accompanist/permissions/MultiplePermissionsState.kt

```diff
@@ -0,0 +1,138 @@
+ * [documentation](https://developer.android.com/training/permissions/requesting#workflow_for_requesting_permissions).
+ *
+ * @param permissions the permissions to control and observe.
+ */
+@Composable
+fun ActivityResultRegistry.rememberMultiplePermissionsState(
+    vararg permissions: String
+): MultiplePermissionsState {
+    val permissionRequested = rememberSaveable { mutableStateOf(false) }
+    val permissionsState = permissions.map {
+        rememberPermissionState(permission = it)
+    }
+    val revokedPermissions = remember {
+        mutableStateOf(permissionsState.filter { !it.hasPermission })
+    }
+    val shouldShowRationale = remember {
+        mutableStateOf(permissionsState.firstOrNull { it.shouldShowRationale } != null)
+    }
+
+    val launcher = rememberActivityResultLauncher(
+        ActivityResultContracts.RequestMultiplePermissions()
+    ) { permissionsResult ->
+        // The permission has been requested
+        permissionRequested.value = true
+        // Keep track of those permissions that were revoked by the user
+        val revokedPermissionsResult = mutableListOf<PermissionState>()
+        for (permission in permissionsResult.keys) {
+            if (permissionsResult[permission] == false) {
+                revokedPermissionsResult.add(permissionsState.first { it.permission == permission })
+            }
+        }
+        // Update shouldShowRationale with the revoked permissions
+        shouldShowRationale.value = revokedPermissionsResult
+            .onEach { it.refreshShouldShowRationale() }
+            .firstOrNull { it.shouldShowRationale } != null
+        // Update revokedPermissions state
+        revokedPermissions.value = revokedPermissionsResult
+    }
+    return remember(launcher) {
+        MultiplePermissionsState(
+            permissions = permissionsState,
+            launcher = launcher,
+            revokedPermissionsState = revokedPermissions,
+            shouldShowRationaleState = shouldShowRationale,
+            permissionRequestedState = permissionRequested
+        )
+    }
+}
+
+/**
+ * A state object that can be hoisted to control and observe multiple permission status changes.
+ *
+ * In most cases, this will be created via [rememberMultiplePermissionsState].
+ *
+ * @param permissions list of permissions to control and observe.
+ * @param launcher [ActivityResultLauncher] to ask for all permissions to the user.
+ * @param revokedPermissionsState [State] with all revoked permissions.
+ * @param shouldShowRationaleState [State] that represents if the user should be presented with a
+ * rationale.
+ * @param permissionRequestedState [State] that represents if the permission has been requested
+ * previously.
+ */
+@Stable
+class MultiplePermissionsState(
+    val permissions: List<PermissionState>,
+    private val launcher: ActivityResultLauncher<Array<String>>,
+    revokedPermissionsState: State<List<PermissionState>>,
```

Comment:
I'm not sure on exposing state parameters here. Could this class create and own the state instead?

Reply:
These parameters are not exposed—they're not variables. They're exposed as _read_ parameters using the property delegate in the class body

Reply:
Well, the public constructor params are `State<T>`. Make the constructor private? 

I personally think the `MultiplePermissionsState` should create and own the individual `State` instances rather than having them passed in.